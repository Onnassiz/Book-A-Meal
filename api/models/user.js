
module.exports = (sequelize, DataTypes) => {
	const User = sequelize.define('user', {
		id: {
			allowNull: false,
			primaryKey: true,
			type: DataTypes.UUID,
			defaultValue: DataTypes.UUIDV4,
		},
		email: {
			type: DataTypes.STRING,
			validate: {
				isEmail: true,
			},
		},
		passwordHash: DataTypes.STRING,
		fullName: DataTypes.STRING,
		role: DataTypes.STRING,
	}, {});
	User.associate = (models) => {
		User.hasMany(models.meal);
		User.hasMany(models.menu);
		User.hasOne(models.profile);
	};
	return User;
};
