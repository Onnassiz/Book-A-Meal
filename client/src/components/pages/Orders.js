import React from 'react';
import OrdersContainer from '../container/OrdersContainer';
import NavBarContainer from '../container/NavBarContainer';

const Orders = () => (
  <div>
    <NavBarContainer page="orders" />
    <OrdersContainer />
  </div>
);

export default Orders;
