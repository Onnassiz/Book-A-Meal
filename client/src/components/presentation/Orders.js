/* eslint react/no-did-update-set-state: 0 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Pagination from 'react-js-pagination';
import { Accordion } from 'react-accessible-accordion';
import OrderAccordion from './partials/OrderAccordion';

class Orders extends Component {
  constructor(props) {
    super(props);
    this.state = {
      orders: [],
      ordersCount: 0,
      activePage: 1,
    };
    this.updateOrdersState = this.updateOrdersState.bind(this);
    this.handlePageChange = this.handlePageChange.bind(this);
  }

  componentDidMount() {
    document.title = 'Orders - Just Eat';
    this.getUserOrders(6, 0);
  }

  getUserOrders(limit, offset, pageNumber = 1) {
    const { getUserOrders } = this.props;
    getUserOrders(limit, offset).then((response) => {
      if (response.status === 200) {
        this.setState({
          orders: response.data.orders,
          ordersCount: response.data.count,
          activePage: pageNumber,
        });
      }
    });
  }

  handlePageChange(pageNumber) {
    const offset = (pageNumber - 1) * 6;
    this.getUserOrders(6, offset, pageNumber);
  }

  updateOrdersState(order) {
    const stateOrders = this.state.orders;
    const index = stateOrders.findIndex(x => x.id === order.id);
    stateOrders[index] = order;
    this.setState({ orders: stateOrders });
  }

  renderOrderAccordion() {
    return (
      <div>
        { this.state.orders.length > 0 ?
          <Accordion>
            {this.state.orders.map(order =>
          (<OrderAccordion
            key={order.id}
            setUpdatedOrder={this.props.setUpdatedOrder}
            updateOrdersState={this.updateOrdersState}
            order={order}
            orders={this.props.orders}
            addArrayToCart={this.props.addArrayToCart}
            user={this.props.user}
            getMealsInOrder={this.props.getMealsInOrder}
          />))}
          </Accordion> :
          <div id="no-menu">
            <h2>You order history is empty</h2>
          </div> }
      </div>
    );
  }

  renderPagination() {
    return (
      <div>{this.state.ordersCount > 4 ?
        <div style={{ textAlign: 'center' }}>
          <Pagination
            hideDisabled
            activePage={this.state.activePage}
            itemsCountPerPage={6}
            totalItemsCount={this.state.ordersCount}
            pageRangeDisplayed={5}
            onChange={this.handlePageChange}
          />
        </div> : ''}
      </div>
    );
  }

  render() {
    return (
      <div id="content-body">
        <h1>Your Orders</h1>
        { this.renderPagination() }
        { this.renderOrderAccordion() }
        { this.renderPagination() }
      </div>
    );
  }
}

Orders.propTypes = {
  getUserOrders: PropTypes.func.isRequired,
  getMealsInOrder: PropTypes.func.isRequired,
  addArrayToCart: PropTypes.func.isRequired,
  setUpdatedOrder: PropTypes.func.isRequired,
  user: PropTypes.object.isRequired,
  orders: PropTypes.object.isRequired,
};

export default Orders;
