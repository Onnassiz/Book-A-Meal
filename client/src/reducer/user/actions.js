import axios from 'axios';
import { setLoading, unsetLoading } from '../formState/actions';

export const SET_USER = 'SET_USER';
export const SET_SIGN_UP_SERVER_ERRORS = 'SET_SIGN_UP_SERVER_ERRORS';
export const SET_SIGN_IN_SERVER_ERRORS = 'SET_SIGN_IN_SERVER_ERRORS';

export function setUser(user) {
  return dispatch => dispatch({
    type: SET_USER,
    id: user.id,
    name: user.name,
    role: user.role,
    email: user.email,
  });
}

function setSignUpErrors(errors) {
  return dispatch => dispatch({
    type: SET_SIGN_UP_SERVER_ERRORS,
    errors,
  });
}

function setSignInErrors(errors) {
  return dispatch => dispatch({
    type: SET_SIGN_IN_SERVER_ERRORS,
    errors,
  });
}

export function postUser(user) {
  return (dispatch) => {
    dispatch(setLoading());
    axios.post('api/v1/auth/signUp', user).then((response) => {
      localStorage.setItem('id', response.data.id);
      localStorage.setItem('email', response.data.email);
      localStorage.setItem('name', response.data.fullName);
      localStorage.setItem('role', response.data.role);
      localStorage.setItem('token', response.data.token);
      setTimeout(() => {
        window.location = '/';
      }, 500);
    }).catch((error) => {
      dispatch(unsetLoading());
      dispatch(setSignUpErrors(error.response.data));
    });
  };
}


export function signInUser(user) {
  return (dispatch) => {
    dispatch(setLoading());
    axios.post('api/v1/auth/signIn', user).then((response) => {
      localStorage.setItem('id', response.data.id);
      localStorage.setItem('email', response.data.email);
      localStorage.setItem('name', response.data.fullName);
      localStorage.setItem('role', response.data.role);
      localStorage.setItem('token', response.data.token);
      setTimeout(() => {
        window.location = '/';
      }, 500);
    }).catch((error) => {
      dispatch(unsetLoading());
      dispatch(setSignInErrors(error.response.data));
    });
  };
}
