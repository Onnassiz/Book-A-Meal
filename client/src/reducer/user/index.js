import { SET_USER, SET_SIGN_UP_SERVER_ERRORS, SET_SIGN_IN_SERVER_ERRORS } from './action';

const initialState = {};

export default (state = initialState, action) => {
	switch (action.type) {
	case SET_USER:
		return { name: action.name, email: action.email, role: action.role };
	case SET_SIGN_UP_SERVER_ERRORS:
		return { signUpErrors: action.errors };
	case SET_SIGN_IN_SERVER_ERRORS:
		return { signInErrors: action.errors };
	default:
		return state;
	}
};
