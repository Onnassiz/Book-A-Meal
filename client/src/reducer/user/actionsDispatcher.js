import axios from 'axios';
import { setLoading, unsetLoading } from '../formState/actionsDispatcher';
import { setUserAction, setSignUpErrorsAction, setSignInErrorsAction } from './actions';

export const setUser = user => dispatch => dispatch(setUserAction(user));

const setSignUpErrors = errors => dispatch => dispatch(setSignUpErrorsAction(errors));
const setSignInErrors = errors => dispatch => dispatch(setSignInErrorsAction(errors));

const processSignIn = (data) => {
  localStorage.setItem('id', data.id);
  localStorage.setItem('email', data.email);
  localStorage.setItem('name', data.fullName);
  localStorage.setItem('role', data.role);
  localStorage.setItem('token', data.token);
  setTimeout(() => {
    window.location = '/';
  }, 500);
};

export const postUser = user => (dispatch) => {
  dispatch(setLoading());
  return axios.post('api/v1/auth/signUp', user).then((response) => {
    dispatch(unsetLoading());
    processSignIn(response.data);
  }).catch((error) => {
    dispatch(setSignUpErrors(error.response.data));
    dispatch(unsetLoading());
  });
};


export const signInUser = user => (dispatch) => {
  dispatch(setLoading());
  return axios.post('api/v1/auth/signIn', user).then((response) => {
    dispatch(unsetLoading());
    processSignIn(response.data);
  }).catch((error) => {
    dispatch(setSignInErrors(error.response.data));
    dispatch(unsetLoading());
  });
};
