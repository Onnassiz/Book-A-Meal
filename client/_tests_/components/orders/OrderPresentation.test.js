import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Orders from '../../../src/components/presentation/Orders';
import user from '../../objectProps/user.props';

const props = {
  getUserOrders: () => new Promise(resolve => resolve({ status: 200 })),
  getMealsInOrder: () => new Promise(resolve => resolve({ status: 200 })),
  addArrayToCart: jest.fn(),
  setUpdatedOrder: jest.fn(),
  orders: {},
  user,
};

const wrapper = shallow(<Orders {...props} />);

describe('<Orders />', () => {
  it('Should render Orders presentation components successfully', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

