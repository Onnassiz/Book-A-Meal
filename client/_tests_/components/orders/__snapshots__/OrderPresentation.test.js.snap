// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Orders /> Should render Orders presentation components successfully 1`] = `
<div
  id="content-body"
>
  <h1>
    Your Orders
  </h1>
  <div />
  <div>
    <div
      id="no-menu"
    >
      <h2>
        You order history is empty
      </h2>
    </div>
  </div>
  <div />
</div>
`;
